File: cpp/cpp_100-999/592.cc
// C++
// leetcode 592
// https://leetcode.cn/problems/fraction-addition-and-subtraction/

class Solution {
public:
	// 计算最大公约数
	int getGcd(int a, int b) {
		if (b > a) {
			swap(a, b);
		}
	    if(b == 0)  return a;
	    return getGcd(b, a % b);
	}

    string fractionAddition(string s) {
    	int arr[20];
    	bool isNegative = false;
    	int tmp = 0, size = 0;

    	// 数据统计
    	for (auto c : s) {
    		if (c == '/') {
    			if (isNegative) {
    				tmp = - tmp;
    				isNegative = false;
    			}
    			arr[size++] = tmp;
    			tmp = 0;
    		} else if (c == '+') {
    			if (isNegative) {
    				tmp = - tmp;
    				isNegative = false;
    			}
    			arr[size++] = tmp;
    			tmp = 0;
    		} else if (c == '-') {
    			if (isNegative) {
    				tmp = - tmp;
    			}
    			if (tmp != 0)
    				arr[size++] = tmp;
    			tmp = 0;
    			isNegative = true;
    		} else {
    			// 数字
    			tmp = 10 * tmp + (c - '0');
    		}
    	}
    	arr[size++] = tmp;

    	tmp = 2;
    	int a = arr[0], b = arr[1], c, d, down;

    	// 处理计算
    	while (tmp < size) {
    		c = arr[tmp++];
    		d = arr[tmp++];
    		down = getGcd(b, d);
    		down = (b / down) * d;
    		a = a * down / b;
    		c = c * down / d;
    		a += c;
    		b = down;
    		down = getGcd(a > 0 ? a : -a, b);
    		a /= down;
    		b /= down;
    	}

    	if (a == 0)
    		b = 1;
    	return to_string(a) + "/" + to_string(b);
    }
};

File: Go/Go_100-999/592.go
import "fmt"

// Go
// leetcode 592
// https://leetcode.cn/problems/fraction-addition-and-subtraction/

// 最大公约数
func getGcd(a, b int) int {
	if b > a {
		a, b = b, a
	}

	if b == 0 {
		return a
	}

	return getGcd(b, a%b)
}

func fractionAddition(expression string) string {
	arr := make([]int, 0, 20)
	isNegative := false
	tmp := 0
	// 数据录入
	for _, c := range expression {
		switch c {
		case '/':
			if isNegative {
				tmp = -tmp
				isNegative = false
			}
			arr = append(arr, tmp)
			tmp = 0
		case '+':
			if isNegative {
				tmp = -tmp
			}
			arr = append(arr, tmp)
			tmp = 0
			isNegative = false
		case '-':
			if isNegative {
				tmp = -tmp
			}
			if tmp != 0 {
				arr = append(arr, tmp)
			}

			tmp = 0
			isNegative = true
		default:
			tmp = tmp*10 + int(c-'0')
		}
	}
	arr = append(arr, tmp)

	// 数据处理
	for len(arr) != 2 {
		// 运算
		tmp = getGcd(arr[1], arr[3])
		tmp = arr[1] * arr[3] / tmp
		arr[0] *= tmp / arr[1]
		arr[2] *= tmp / arr[3]
		arr[2] += arr[0]
		arr[3] = tmp
		tmp = arr[2]

		// 化简
		if tmp < 0 {
			tmp = -tmp
		}
		tmp = getGcd(tmp, arr[3])
		arr[2] /= tmp
		arr[3] /= tmp

		// 去掉头两位
		arr = arr[2:]
	}

	res := fmt.Sprint(arr[0], "/", arr[1])
	return res
}